components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java

```diff
@@ -0,0 +1,174 @@
+
+/**
+ * This class holds the implementation of the CertificateManager interface.
+ */
+public class CertificateManagerImpl implements CertificateManager {
+
+    private static Log log = LogFactory.getLog(CertificateManagerImpl.class);
+    private static final String PROFILE_CONFIG = "sslprofiles.xml";
+    private static String CARBON_HOME = System.getProperty("carbon.home");
+    private static final char SEP = File.separatorChar;
+    private static String SSL_PROFILE_FILE_PATH = CARBON_HOME + SEP + "repository" + SEP + "resources" + SEP
+            + "security" + SEP + PROFILE_CONFIG;
+    private static CertificateMgtDAO certificateMgtDAO = CertificateMgtDAO.getInstance();
+    private static CertificateMgtUtils certificateMgtUtils = new CertificateMgtUtils();
+
+    @Override
+    public ResponseCode addCertificateToPublisher(String certificate, String alias, String endpoint, int tenantId) {
+        try {
+            if (certificateMgtDAO.addCertificate(alias, endpoint, tenantId)) {
+                ResponseCode responseCode = certificateMgtUtils.addCertificateToTrustStore(certificate, alias);
+                if (responseCode.getResponseCode() ==
+                        ResponseCode.INTERNAL_SERVER_ERROR.getResponseCode()) {
+                    log.error("Error adding the certificate to Publisher Trust Store. Rolling back...");
+                    certificateMgtDAO.deleteCertificate(alias, endpoint, tenantId);
+                } else if (responseCode.getResponseCode() == ResponseCode.ALIAS_EXISTS_IN_TRUST_STORE
+                        .getResponseCode()) {
+                    log.error("Could not add Certificate to Trust Store. Certificate Exists. Rolling back...");
+                    certificateMgtDAO.deleteCertificate(alias, endpoint, tenantId);
+                } else if (responseCode.getResponseCode() == ResponseCode.CERTIFICATE_EXPIRED.getResponseCode()) {
+                    log.error("Could not add Certificate. Certificate expired.");
+                    certificateMgtDAO.deleteCertificate(alias, endpoint, tenantId);
+                } else {
+                    log.info("Certificate is successfully added to the Publisher client Trust Store with Alias '"
+                            + alias + "'");
+                }
+                return responseCode;
+            } else {
+                log.error("Error persisting the certificate meta data in db. Certificate could not be added to " +
+                        "publisher Trust Store.");
+                return ResponseCode.INTERNAL_SERVER_ERROR;
+            }
+        } catch (CertificateManagementException e) {
+            String cause = e.getMessage();
+            return cause.contains("Alias") ? ResponseCode.ALIAS_EXISTS_IN_TRUST_STORE : ResponseCode
+                    .CERTIFICATE_FOR_ENDPOINT_EXISTS;
+        }
+    }
+
+    @Override
+    public ResponseCode deleteCertificateFromPublisher(String alias, String endpoint, int tenantId) {
```

Comment:
better to avoid response code from impl layer., I believe this is like http status codes?

Reply:
The reason to returning response codes from here is, as certificate addition can be failed in multiple levels, we need to know where it fails and show the error in ui.